### PR TITLE
#437 to add Role and Permission entities to a module

### PR DESCRIPTION
**Pull request by masumcse1**
Problem 1. Following liquibase changlog file need to execute before meveo deployment in wildfy , but 
  this file was not executed  before meveo deployment. So, it makes error during deployment time.
 
 meveo-model/src/main/db_resources/changelog/current/structure.xml
 
 problem 2 : need to set some  value for "adm_role" table .I can not set those value. 
 
 Example: 
  update adm_role set creator='meveo.admin' , updater='meveo.admin'

 how can i represent those  value in structure.xml according to meveo development guideline ?
 
Another example  : 
   update adm_role set code=role_name , description=role_description 
  
  for unit testing purpose i was putted some data and it works fine.
  But How can i put real data for those two columns (code , description) or any importance to put real data 
  into table for this case ?

### CODE REVIEW

**Comment by ClementBareth** [comment]
Problem 1 : the changeds needs to be append to the 7.0-structure.xml file. Then you just need to run liquibase:update -Padditional to execute them

**Comment by ClementBareth** [comment]
The example values you provided are fine

File: meveo-api/src/main/java/org/meveo/api/PermissionApi.java [hunk 1]
```diff
@@ -5,28 +5,91 @@
...
+				PermissionDto pd = new PermissionDto(p);
+				permissionsDto.getPermission().add(pd);
+			}
+		}
+
+		return permissionsDto;
+	}
+
+	@Override
+	public PermissionDto find(String code) throws EntityDoesNotExistsException, MissingParameterException
```
Comment (ClementBareth) [inline meveo-api/src/main/java/org/meveo/api/PermissionApi.java]: This needs to be implemented

File: meveo-api/src/main/java/org/meveo/api/PermissionApi.java [hunk 2]
```diff
@@ -5,28 +5,91 @@
...
+	        throw new EntityDoesNotExistsException(Role.class, name, "name");
+	    }
+	    roleDto = new RoleDto(role, true, true);
+	
+	    return roleDto;*/
+	return null;
+	}
+
+	@Override
+	public Permission createOrUpdate(PermissionDto dtoData) throws MeveoApiException, BusinessException {
```
Comment (ClementBareth) [inline meveo-api/src/main/java/org/meveo/api/PermissionApi.java:69]: This needs to be implemented

File: meveo-model/src/main/db_resources/changelog/current/structure.xml [hunk 3]
```diff
@@ -1298,4 +1298,67 @@
 			<where>id IN (-1,-2,-3,-4,-5,-6,-7,-8,-9,-10,-11,-12,-13,-14,-15)</where>
 		</delete>
 	</changeSet>
+	
+	<changeSet author="Gulam Samdani" id="#437 -Convert to buss entity for adm_permission">
```
Comment (ClementBareth) [inline meveo-model/src/main/db_resources/changelog/current/structure.xml]: Move the changeset to 7.0-structure.xml

**Review by ClementBareth — CHANGES_REQUESTED** [review]

**Comment by ClementBareth** [comment]
@masumcse1   In order to have the possibility to set the values, you need to add the not null constraint after setting the data, in another instruction. To set dynamically the data, you can use <sql> tag to write plain sql

**Comment by masumcse1** [comment]
problem1: PermissionApi.java

How many class attribute  need to consider for this permission entity creation method?
I have considered only two field : name , permission currently  where as this entity has several field.
I have added few attribute for this entity class .

problem2: PermissionApi.java

I have created two method (createOrUpdate,find) for Permission entity in this class . 
is there any necessary to implement those two function  in other class (PermissionRsImpl,PermissionRs) .

problem 3 : PermissionApi.java
I was not done any unit test for those two (createOrUpdate,find) method.
How can i do unit test for those two newly created method ?

problem3 : What is the request execution flow for bellow class ?

Example : 
....Api.java 
what is the  sequence are using for calling this class's method ? 

I have confusion but i seems that  sequence of function calling system is like

RS > WsImpl> API >API/ Service
JSFBean>API/Service>Service

**Comment by ClementBareth** [comment]
Problem 1 => You can set name = description, and permission = code for now
Problem 2 => No, it's ok
Problem 3 => If you go to src/test/apiTests/postman/tests there are a lot of postman test suites that are used as unit /integration testing at the API level
Problem 4:
*  The REST sequence is RS >  API > API/ Service
* The JSF Sequence is correct

File: meveo-model/src/main/db_resources/changelog/releases/7.0-structure.xml [hunk 4]
```diff
@@ -44,10 +44,76 @@
...
+					</addColumn>
+					<addColumn tableName="adm_role">
+						<column name="updated" type="datetime" defaultValueDate="CURRENT_TIMESTAMP"/>
+					</addColumn>
+					<addColumn tableName="adm_role">
+						<column name="code" type="varchar(255)"/>
+					</addColumn>
+					<addColumn tableName="adm_role">
+						<column name="description" type="varchar(255)"/>
+					</addColumn>
```
Comment (ClementBareth) [inline meveo-model/src/main/db_resources/changelog/releases/7.0-structure.xml:114]: Here you miss following points : 

* Update of data in the table (for role : code = name, for permission, code = permission & description = name)
* Addition of unique / not null constraint on code

**Review by ClementBareth — CHANGES_REQUESTED** [review]

**Comment by masumcse1** [comment]
mvn liquibase:update -Padditional  

above command is not working  for meveo model.

image

File: meveo-api/src/main/java/org/meveo/api/PermissionApi.java [hunk 5]
```diff
@@ -5,28 +5,184 @@
...
+
+		if (!(currentUser.hasRole("superAdminManagement") || (currentUser.hasRole("administrationManagement")))) {
+			throw new ActionForbiddenException("User has no permission to manage permissions.");
+		}
+
+		Permission permission = new Permission();
+		permission.setName(postData.getDescription());
+		permission.setPermission(postData.getCode());
+		
+		permissionService.create(permission);
```
Comment (ClementBareth) [inline meveo-api/src/main/java/org/meveo/api/PermissionApi.java:131]: Code & Description fields of the Permission entity also need to be filled

File: meveo-api/src/main/java/org/meveo/api/PermissionApi.java [hunk 6]
```diff
@@ -5,28 +5,184 @@
...
+		}
+
+		handleMissingParameters();
+
+		if (!(currentUser.hasRole("superAdminManagement") || (currentUser.hasRole("administrationVisualization")))) {
+			throw new ActionForbiddenException("User has no permission to access roles");
+		}
+
+		PermissionDto permissionDto = null;
+		Permission permissionEntity = permissionService.findByPermission(permission);
```
Comment (ClementBareth) [inline meveo-api/src/main/java/org/meveo/api/PermissionApi.java]: Here use the code to retrieve the permission

File: meveo-api/src/main/java/org/meveo/api/PermissionApi.java [hunk 7]
```diff
@@ -5,28 +5,184 @@
...
+		}
+
+		if (postData.getName() != null) {
+			permission.setName(postData.getDescription());
+		}
+
+		if (postData.getPermission() != null) {
+			permission.setPermission(postData.getCode());
+		}
+
```
Comment (ClementBareth) [inline meveo-api/src/main/java/org/meveo/api/PermissionApi.java]: Same remarks as create

File: meveo-model/src/main/db_resources/changelog/releases/7.0-structure.xml [hunk 8]
```diff
@@ -44,10 +44,90 @@
...
+				
+		</changeSet>
+		
+		<changeSet author="Gulam Samdani" id="#437 update data for adm_role and adm_permission">
+		   <sql> update adm_role set code=role_name ,description=role_description </sql>
+		   <sql> update adm_permission  set code=permission  ,description=name </sql> 
+        </changeSet>
+   	
+
```
Comment (ClementBareth) [inline meveo-model/src/main/db_resources/changelog/releases/7.0-structure.xml]: After the update you still need to add the constraints on code columns

**Review by ClementBareth — CHANGES_REQUESTED** [review]

**Comment by masumcse1** [comment]
i have created a new method "findByCode' in PermissionService. 

Please review my code.

File: meveo-model/src/main/db_resources/changelog/releases/7.0-structure.xml [hunk 9]
```diff
@@ -44,10 +44,92 @@
...
+		</changeSet>
+		
+		<changeSet author="Gulam Samdani" id="#437 update data for adm_role and adm_permission">
+		   <sql> update adm_role set code=role_name ,description=role_description </sql>
+		   <sql> update adm_permission  set code=permission  ,description=name </sql> 
+        </changeSet>
+   	
+   	 <changeSet author="Gulam Samdani" id="#437 set not null constraint for code column in adm_permission ">
+		   <sql>  ALTER TABLE adm_permission ALTER COLUMN code SET NOT NULL ; </sql> 
+	 </changeSet>
```
Comment (ClementBareth) [inline meveo-model/src/main/db_resources/changelog/releases/7.0-structure.xml:126]: adm_role (code) should also be not null

**Review by ClementBareth — CHANGES_REQUESTED** [review]

File: meveo-model/src/main/db_resources/changelog/releases/7.0-structure.xml (hunk 9 above)
Comment (masumcse1) [reply meveo-model/src/main/db_resources/changelog/releases/7.0-structure.xml:126]: issue resolved .please check this code.

**Review by ClementBareth — APPROVED** [review]

**Comment by ClementBareth** [comment]
@masumcse1 Can you merge master into your branch in order to resolve conflicts ?

**Comment by masumcse1** [comment]
Git  merge completed for this  branch.

**Pull request closed without merging**